### DISCUSSION TITLE
Linear elasticity using periodic boundary conditions

### DISCUSSION THREAD

**Discussion by dthillaithevan** [firedrakeproject/firedrake#2661] [Firedrake support]
Hey, 

I have been attempting to use PeriodicSquareMesh to solve a linear elastic problem to compute displacments. My goal is to impose periodic boundary conditions (PBCs) on every exterior surface, apply global strains and then compute the resulting displacements (which should be periodic across every surface). 

With `direction='both'` the solution is appears incorrect with a constant displacement throughout the domain. However, with `direction='x'` or `direction='y'` the displacements seem more reasonable (please see attached screenshots of displacements). However, I would like to impose the PBCs on every exterior surface so I believe I need to use the `direction='both'` option, does anyone have any ideas as to what could be going wrong here? Is there any reason to expect the mathematical formulation to differ from FEniCS?

Thank you! 

```
from firedrake import *

mesh = PeriodicSquareMesh(10,10,1, direction='both', quadrilateral = False) #Constant displacements when direction = 'both'
V = VectorFunctionSpace(mesh, "CG", 1)
U = FunctionSpace(mesh, 'CG', 1)
u = TrialFunction(V)
v = TestFunction(V)# Identity tensor
Id = Identity(mesh.geometric_dimension())
lambda_ = Constant(0.25)
mu = Constant(1)

def epsilon(a):
    return sym(grad(a))

def sigma(a,lmbda_func,mu_func,Eps):
    return lmbda_func*tr(epsilon(a) + Eps)*Id + 2*mu_func*(epsilon(a)+Eps) #Impose additional global strain, Eps

Eps = as_vector([[0,0.5],[0.5,0]]) # Apply unit shear strain
dv = TrialFunction(V)
v_ = TestFunction(V)
v = Function(V)

F = inner(sigma(dv,lambda_,mu, Eps), epsilon(v_))*dx
a = lhs(F)
L = rhs(F)
problem = LinearVariationalProblem(a, L, v, [])
solver = LinearVariationalSolver(problem)
solver.solve()

print (v.vector()[:]) #Constant when direction='both'
```
`direction='y'` solution
![direction_y_solution](https://user-images.githubusercontent.com/44264236/205017205-8c8e355a-629d-4910-9a60-580da00bfa19.png)

`direction='x'` solution
![direction_x_solution](https://user-images.githubusercontent.com/44264236/205017215-f4f7d29e-ea90-4be7-84cc-48651b314296.png)

`direction='both'` solution
![direction_both_solution](https://user-images.githubusercontent.com/44264236/205017228-b2dec74e-8d9d-41fd-a0fa-4ddf10442e94.png)

**Comment by dham** [comment]
I think your problem is ill-posed. The elasticity operator with no Dirichlet conditions has a non-trivial nullspace which you'd need to remove. See: https://firedrakeproject.org/solving-interface.html#solving-singular-systems

**Reply by dthillaithevan** [reply]
Thanks for your quick reply David, I was not aware of this problem. 

As a initial starting point I assumed the nullspaces to be the 2 translation vectors and the 1 rotation vector, but this does not seem to be correct. Do you have any suggestions or advice for determining the correct nullspace?  

```
x, y = SpatialCoordinate(mesh)
b0.interpolate(Constant([1, 0]))
b1.interpolate(Constant([0, 1]))
b2.interpolate(as_vector([-y, x]))
nullspace = VectorSpaceBasis([b0, b1, b2])
# nullspace = VectorSpaceBasis(constant=True)
nullspace.orthonormalize()
Eps = as_vector([[0,0.5],[0.5,0]]) # Apply unit x strain
dv = TrialFunction(V)
v_ = TestFunction(V)
v = Function(V)
F = inner(sigma(dv,lambda_,mu, Eps), epsilon(v_))*dx
a = lhs(F)
L = rhs(F)
problem = LinearVariationalProblem(a, L, v, [])
solver = LinearVariationalSolver(problem, nullspace= nullspace)
solver.solve()
```

**Comment by ksagiyam** [comment]
Also, I think your residual strain `Eps` is incompatible with the doubly periodic boundary. Please note that even if strain is periodic, displacement in general is not periodic: we have, e.g., `u(y=1) = u(y=0) + cy` where `cy` is some vector constant; consider the following simple shear case:
```
  ______
 /     /
/_____/
```
In the above simple example, you might be able to map your problem as `ubar = u - cy * y` and solve for `ubar` on the periodic mesh, though.

**Reply by dthillaithevan** [reply]
Thanks ksagiyam! 

Yes I do agree with you. I should have been more clear above, I am solving for the periodic fluctuation here (v). To get the total displacement from v I would do something like: `0.5*(dot(Eps, SpatialCoordinate(mesh))+v)`, which is the equation you are referring to I think.

I have been playing around with the code and my main confusion is still why there is a disparity between the solutions when setting `direction='x' or 'y'` compared to `direction='both'`. With the nullspace defined as above, setting `direction='y'` and applying a shear strain I am able to get periodic results which are more aligned to what I would expect. The displacement along y (v_y) are as you would expect but v_x is incorrect, although still periodic. However, setting `direction='both'` still gives 0 periodic displacements. Could this be related to nullspaces? Is there anyway to check the nullspace is defined correctly?  


v_y with `direction='y'`
![Screenshot 2022-12-01 at 10 53 28](https://user-images.githubusercontent.com/44264236/205034950-a7a87986-f460-490e-9f99-73226010b0fb.png)
v_x with `direction='y'`
![Screenshot 2022-12-01 at 10 53 45](https://user-images.githubusercontent.com/44264236/205035071-b2841c3d-7a09-42bb-953d-04d0ed8c5cce.png)

**Reply by ksagiyam** [reply]
If what you posted was the code for the mapped problem, I think your definition of `sigma` is not correct; it looks to me the definition of stress in the unmapped (original) problem. Specifically, for doubly periodic problem (`both`), I think `Eps` should vanish in the definition of `sigma`, which then gives you `v = 0` after solving with appropriate treatment of the nullspace, from which you can reconstruct the solution to the original problem as `v_orig = v + Eps x` where `v = 0` and `x` is the coordinate vector, and you can check that `epsilon(v_orig) = Eps`, which I believe is what you expect.

**Reply by dthillaithevan** [reply]
Thanks for this! I am not sure I fully understand. What do you mean by original problem and mapped problem?

Also FYI, the code I posted above is a modified version of the [PBC tutorial for FEniCS](https://comet-fenics.readthedocs.io/en/latest/demo/periodic_homog_elas/periodic_homog_elas.html)

**Reply by wence-** [reply]
Add `"ksp_test_null_space": None` to your solver parameters to test the nullspace at the discrete level.

Note that if you are periodic in both directions the null space is both translational modes (but I think not the rotations), whereas if you are only periodic in one direction and are applying dirichlet conditions on the other side your null space is only the translational mode in the periodic direction.

**Reply by dthillaithevan** [reply]
Thanks! 
Yes! Getting rid of the rotational mode appears to improve the solution for `direction='both'`. I am getting v ≈ 0 as suggested by @ksagiyam, however the `v_orig = v + Eps x`solution displays weird behaviour at the vertices of the domain. Not sure if this is caused the underlying peridocity mapping being incorrect? I used to have similar errors in FEniCS when the PBC dof mapping was incorrect. 

@ksagiyam - I am not entirely sure I fully follow why we require `v = 0` as the solution. When `v = 0` then `v_orig = Eps x`, and `Eps` is given, so what are we solving for in this case? 

![v_orig_solution_for_shear](https://user-images.githubusercontent.com/44264236/205302832-7d351638-6861-4907-a019-363bdf8ba73d.png)

I also added `"ksp_test_null_space": None` but didnt seem to produce any additional outputs, do I need to add any additional options? Or is a lack of additional output a good sign (no errors)?

**Reply by ksagiyam** [reply]
My understanding is that you are (at least in the code you provided) trying to solve a linear elasticity problem for the displacement field `u`, where the material is under prescribed periodic strain, which you denote by `Eps`, and you want to solve this problem using a periodic mesh. If you choose `u` as the primary variable, you cannot define `u` directly on the function space defined on the periodic mesh as `u` is not periodic in general, but if you define `ubar` so that `u = \int Eps\dot dx + ubar` (or `u = Eps\dot x + ubar` if Eps is constant), you can define `ubar` on the function space defined on the periodic mesh and solve the linear elasticity problem (rewritten in terms of `ubar`). If `Eps` is as simple as that defined in your code, we know that `ubar = 0`. If `Eps` was something more interesting, `ubar` would be someting more interesting.

I also observed that `ksp_test_null_space` did not say anything even if I set completely wrong nullspaces; I'm not sure why.

The original displacement field is not periodic, so you would need to use DG space as in:
```
DG = FunctionSpace(mesh, "DG", 1)
v_orig = Function(DG).interpolate(v + Eps x)
```

**Reply by stephankramer** [reply]
I think you need `"mat_null_space_test_view": None` in addition to `"ksp_test_null_space": None` for it to print the result to the log.

**Reply by dthillaithevan** [reply]
> My understanding is that you are (at least in the code you provided) trying to solve a linear elasticity problem for the displacement field `u`, where the material is under prescribed periodic strain, which you denote by `Eps`, and you want to solve this problem using a periodic mesh. If you choose `u` as the primary variable, you cannot define `u` directly on the function space defined on the periodic mesh as `u` is not periodic in general, but if you define `ubar` so that `u = \int Eps\dot dx + ubar` (or `u = Eps\dot x + ubar` if Eps is constant), you can define `ubar` on the function space defined on the periodic mesh and solve the linear elasticity problem (rewritten in terms of `ubar`). If `Eps` is as simple as that defined in your code, we know that `ubar = 0`. If `Eps` was something more interesting, `ubar` would be someting more interesting.
> 
> I also observed that `ksp_test_null_space` did not say anything even if I set completely wrong nullspaces; I'm not sure why.
> 
> The original displacement field is not periodic, so you would need to use DG space as in:
> 
> ```
> DG = FunctionSpace(mesh, "DG", 1)
> v_orig = Function(DG).interpolate(v + Eps x)
> ```

Hi @ksagiyam,

I completely agree with what you have said, although my strain is not periodic (it is a global macrostrain) the rest of your statements are correct. This was a test problem I had setup to get my head around periodicity in firedrake, my end goal is to modify the material distribution throughout the domain which should result in more interesting displacement distributions. Thanks a lot for your help!

And thank you, @stephankramer, adding that option does indeed print the results! 

These are the main outputs which suggests my nullspace is incorrect. I will try to determine the correct vectors.  
`Null vector 0 unlikely null vector || A * v[0] || = 5.3605e-05`
`Null vector 1 unlikely null vector || A * v[1] || = 4.70992e-05`

**Reply by dthillaithevan** [reply]
Hey again,

Am I right to assume that another way to remove the rigid body translations is to introduce Lagrange multiplers? That is, to enforce a 0 average of the perodic displacement `ubar` (which is equivalent to constraining a single node in the mesh).

And if so, this requires an additional `R` space to be included in the formulation. However when I do this I get an error: `UFL:ERROR Replacement expressions must have the same shape as what they replace.` 



```
from firedrake import *

n = 20
mesh = PeriodicSquareMesh(n,n,1, direction='both', quadrilateral = True)

V = VectorFunctionSpace(mesh, "CG", 1)
R = VectorFunctionSpace(mesh, 'R', 0) # Real space for Lagrange
W = V * R
w = Function(W)
U = FunctionSpace(mesh, 'CG', 1)
dv, dlamb = TrialFunctions(W)
v_, lamb_ = TestFunctions(W)

# Identity tensor
Id = Identity(mesh.geometric_dimension())

E = 209e9
nu = 0.3
lambda_ = Constant(E*nu/(1+nu)/(1-2*nu))
mu = Constant(E/2/(1+nu))

def epsilon(a):
    return sym(grad(a))
    
def sigma(a,lmbda_func,mu_func,Eps):
    return lmbda_func*tr(epsilon(a) + Eps)*Id + 2*mu_func*(epsilon(a)+Eps)

E_func = Function(U)
lmbda_func = Function(U)
mu_func = Function(U)
# Circular inclusion to introduce material variablility
E_func.interpolate(conditional((x-0.5)**2 + (y-0.5)**2 < 0.1, 210e3, 50e3))

lmbda_func.vector()[:] = E_func.vector()[:]*nu/ ((1+nu)*(1-2*nu))
mu_func.vector()[:] = E_func.vector()[:] / (2*(1+nu))

Eps = as_vector([[0,0.5],[0.5,0]]) # Apply shear strain

F = inner(sigma(dv,lmbda_func,mu_func, Eps), epsilon(v_))*dx
a = lhs(F)
L = rhs(F)
a += dot(lamb_,dv)*dx + dot(dlamb,v_)*dx
# Taken from https://github.com/firedrakeproject/firedrake/issues/2664
solver_parameters = {
    "ksp_type": "fgmres",
    "mat_type": "matfree",
    "pc_type": "fieldsplit",
    "pc_fieldsplit_type": "schur",
    "pc_fieldsplit_schur_fact_type": "full",
    "pc_fieldsplit_0_fields": "0,1",
    "pc_fieldsplit_1_fields": "2",
    "fieldsplit_1_ksp_type": "gmres",
    "fieldsplit_1_ksp_max_it": 1,
    "fieldsplit_1_pc_type": "none",
    "fieldsplit_0": {
        "ksp_atol": 1e-12,
        "ksp_rtol": 1e-12,
        "ksp_type": "preonly",
        "pc_type": "python",
        "pc_python_type": "firedrake.AssembledPC",
        "assembled_pc_type": "lu",
        "assembled_pc_factor_mat_solver_type": "mumps"
    },
}

solve(a == L, w, [], solver_parameters = solver_parameters)
```

**Reply by ksagiyam** [reply]
`VectorFunctionSpace(mesh, 'R', 0)` seems not fully supported; please see https://github.com/firedrakeproject/firedrake/issues/1456. 

I think you could do something like the following:
```
...
R = FunctionSpace(mesh, 'R', 0)
W = V * R * R
w = Function(W)
U = FunctionSpace(mesh, 'CG', 1)
dv, dlamb0, dlamb1 = TrialFunctions(W)
v_, lamb_0, lamb_1 = TestFunctions(W)
dlamb = as_vector([dlamb0, dlamb1])
lamb_ = as_vector([lamb_0, lamb_1])
...
# The rest of the code is unchanged.
```
but I don't think this is the best thing to do. Could you tell us why you do not want the solver to remove the rigid body modes?

**Reply by dthillaithevan** [reply]
Thank you Koki, that worked!

It was mainly to test my understanding. I wanted to check that the solutions obtained by removing the rigid body modes with nullspaces was equivalent to the Lagrangian approach. Which, using your code above, I think is the case. 

I am however still having some weird total displacement (`v_orig`) outputs where the displacement at the vertices of the domain appears to be incorrect. 

I have tried to project into `DG` but I get the following error: `RuntimeError: Expression of length 1 required, got length 2`. And I have tried using higher order meshes and projecting into low order e.g. `VectorFunctionSpace(mesh, "CG", 2)` and projecting to `VectorFunctionSpace(mesh, "CG", 1)`, but still get the same results (please see screenshot attached). 

Do you know what could be causing this?

![image](https://user-images.githubusercontent.com/44264236/207591218-85638146-4b95-4262-bfa7-1c798a879438.png)

**Reply by ksagiyam** [reply]
I think I forgot to say `Vector` in the earlier suggestion (and I also forgot to include `variant`); you would need to interpolate your displacement into:
```
DG = VectorFunctionSpace(mesh, "DG", 1, variant="equispaced")
```
This `variant` argument is to include end points. The default `variant` is `spectral`, which would use Gauss-Legendre nodes and would not include end points.

**Reply by dthillaithevan** [reply]
Thanks! 
I get the following error when I try that - `VectorFunctionSpace() got an unexpected keyword argument 'variant'`

Could it be to do with my firedrake version?

**Reply by ksagiyam** [reply]
Sorry, we need to do this at lower level:
```
import ufl

elem = ufl.VectorElement("Q", mesh.ufl_cell(), 1, variant="equispaced")
DG = FunctionSpace(mesh, elem)
```

**Reply by dthillaithevan** [reply]
Thanks for your quick reply!

The code works but the output remains the same, do you think there could be something else going on here?

![image](https://user-images.githubusercontent.com/44264236/207611309-bbc939db-ad11-45eb-b309-af0145d48909.png)

**Reply by ksagiyam** [reply]
Could you tell us how you plot the displacement field? If you directly look at the DoF values, do they look sensible? (You might want to make your problem as small as possible to check the DoF values.)

**Reply by dthillaithevan** [reply]
Hi Koki,

I am using paraview and saving the projected function to a `.pvd` file.

I've had a look at the individual components of `v_orig = v + Eps x` and it seems the problem is `Eps x`. 

Below are screenshots of the displacements `v_orig`, `v` and `Eps x` (in order) and you can see that `v` looks reasonable but `Eps x` introduces the odd behaviour at the vertices which then results in `v_orig` having similar results.

`v_orig`
![image](https://user-images.githubusercontent.com/44264236/208056739-4d589dd3-2cd7-43e7-9fd9-3df3f6f5ee2c.png)
`v`
![image](https://user-images.githubusercontent.com/44264236/208056652-ec430df5-e2ba-4f6d-82d7-7a591b0980f8.png)

`Eps x`
![image](https://user-images.githubusercontent.com/44264236/208056582-b2a7d4f6-86f8-413b-a931-a37cdfec00f8.png)

Do you know why this could be the case?

**Reply by ksagiyam** [reply]
`v` looks good. It's hard to say without looking at the code. Could you make a minimal example that shows this weird behaviour? You should only need to define `mesh`, `x`, and `Eps`.

**Reply by dthillaithevan** [reply]
Sure, here you go.

```
import ufl
from firedrake import *

n = 20
mesh = PeriodicSquareMesh(n,n,1, direction='both', quadrilateral = True)
CG = VectorFunctionSpace(mesh, "CG", 1)
elem = ufl.VectorElement("Q", mesh.ufl_cell(), 1, variant="equispaced")
DG = FunctionSpace(mesh, elem)

Eps = as_vector([[0,0.5],[0.5,0]])

outfile = File("DG_macro_EPS.pvd")
outfile.write(project((dot(Eps, SpatialCoordinate(mesh))), DG))

outfile = File("CG_macro_EPS.pvd")
outfile.write(project((dot(Eps, SpatialCoordinate(mesh))), CG))
```

**Reply by ksagiyam** [reply]
Sorry again, please use "DQ" instead of "Q".

**Reply by dthillaithevan** [reply]
That works, thanks a lot for your help Koki, appreciate it!

**Reply by dthillaithevan** [reply]
> Sorry again, please use "DQ" instead of "Q".

Hi Again Koki,

Is there an extension of this code to 3D? 
```
elem = ufl.VectorElement("Q", mesh.ufl_cell(), 1, variant="equispaced")
DG = FunctionSpace(mesh, elem) 
```

At the moment I get the following error when I run this code on a 3D periodic mesh:

```
  File "/Users/dee/firedrake/src/ufl/ufl/log.py", line 158, in error
    raise self._exception_type(self._format_raw(*message))

UFLException: Cellname "tetrahedron" invalid for "DQ" finite element.
```

Thanks!

**Reply by ksagiyam** [reply]
`PeriodicCubeMesh` and friends currently do not support hex cells and so use tetrahedral cells, in which case you want "DP" (or "DG") instead of "DQ".
